File: frontend/src/app/shared/proof-of-work.ts
/**
 * Proof-of-work řešič pro formulář zpětné vazby (docs/nasazeni.md, obrana proti spamu) —
 * náhrada CAPTCHY, kterou appka nesmí použít (docs/soukromi.md, žádné externí skripty/CDN
 * třetí strany). Definice musí být BIT-PŘESNĚ stejná jako backend
 * (`FeedbackChallengeService`) a mobil (`ProofOfWork.kt`): hledá se nejmenší nezáporné celé
 * `nonce`, pro které má SHA-256(UTF8(salt + ":" + nonce)) `difficulty` vedoucích nulových bitů.
 *
 * Čistá funkce nad Web Crypto, testovatelná bez Workeru — samotné volání v appce jde přes
 * `proof-of-work.worker.ts`, ať se počítání neděje na hlavním vlákně.
 */
export async function solveProofOfWork(
  salt: string,
  difficulty: number,
  signal?: AbortSignal,
): Promise<string> {
  const encoder = new TextEncoder();
  for (let nonce = 0; ; nonce++) {
    if (signal?.aborted) {
      throw new DOMException('Proof-of-work zrušeno', 'AbortError');
    }
    const data = encoder.encode(`${salt}:${nonce}`);
    const hashBuffer = await crypto.subtle.digest('SHA-256', data);
    if (hasLeadingZeroBits(new Uint8Array(hashBuffer), difficulty)) {
      return String(nonce);
    }
  }
}

export function hasLeadingZeroBits(hash: Uint8Array, bits: number): boolean {
  const fullBytes = Math.floor(bits / 8);
  for (let i = 0; i < fullBytes; i++) {
    if (hash[i] !== 0) return false;
  }
  const remainingBits = bits % 8;
  if (remainingBits === 0) return true;
  const mask = (0xff << (8 - remainingBits)) & 0xff;
  return (hash[fullBytes] & mask) === 0;
}

File: mobile/app/src/main/java/cz/kvalitacena/ui/feedback/ProofOfWork.kt
package cz.kvalitacena.ui.feedback

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.security.MessageDigest

/**
 * Proof-of-work řešič pro formulář zpětné vazby (docs/nasazeni.md, obrana proti spamu) — náhrada
 * CAPTCHY, kterou appka nesmí použít (docs/soukromi.md, žádné externí skripty/CDN třetí strany).
 * Definice musí být BIT-PŘESNĚ stejná jako backend (`FeedbackChallengeService`) a web
 * (`shared/proof-of-work.ts`): hledá se nejmenší nezáporné celé `nonce`, pro které má
 * SHA-256(UTF8(salt + ":" + nonce)) [difficulty] vedoucích nulových bitů.
 */
object ProofOfWork {

  suspend fun solve(salt: String, difficulty: Int): String = withContext(Dispatchers.Default) {
    computeNonce(salt, difficulty)
  }

  private fun computeNonce(salt: String, difficulty: Int): String {
    val digest = MessageDigest.getInstance("SHA-256")
    var nonce = 0L
    while (true) {
      val hash = digest.digest("$salt:$nonce".toByteArray(Charsets.UTF_8))
      if (hasLeadingZeroBits(hash, difficulty)) return nonce.toString()
      nonce++
    }
  }

  fun hasLeadingZeroBits(hash: ByteArray, bits: Int): Boolean {
    val fullBytes = bits / 8
    for (i in 0 until fullBytes) {
      if (hash[i].toInt() != 0) return false
    }
    val remainingBits = bits % 8
    if (remainingBits == 0 || fullBytes >= hash.size) return true
    val mask = (0xFF shl (8 - remainingBits)) and 0xFF
    return (hash[fullBytes].toInt() and mask) == 0
  }
}
